oauthbearer.go

```diff
@@ -61,3 +64,121 @@ func (a *oauthBearerClient) Next(challenge []byte) ([]byte, error) {
 func NewOAuthBearerClient(opt *OAuthBearerOptions) Client {
 	return &oauthBearerClient{*opt}
 }
+
+type OAuthBearerAuthenticator func(opts OAuthBearerOptions) *OAuthBearerError
+
+type oauthBearerServer struct {
+	done         bool
+	fail         error
+	authenticate OAuthBearerAuthenticator
+}
+
+func (a *oauthBearerServer) Next(response []byte) (challenge []byte, done bool, err error) {
+	// Per RFC, we cannot just send an error, we need to return JSON-structured
+	// value as a challenge and then after getting dummy response from the
+	// client stop the exchange.
+	if a.fail != nil {
+		if len(response) != 1 && response[0] != 0x01 {
+			return nil, true, errors.New("unexpected response")
+		}
+		return nil, true, a.fail
+	}
+	fail := func(descr string) ([]byte, bool, error) {
+		blob, err := json.Marshal(OAuthBearerError{
+			Status:  "invalid_request",
+			Schemes: "bearer",
+		})
+		if err != nil {
+			panic(err) // wtf
+		}
+		a.fail = errors.New(descr)
+		return blob, false, nil
+	}
+
+	if a.done {
+		err = ErrUnexpectedClientResponse
+		return
+	}
+	if response == nil {
+		return []byte{}, false, nil
+	}
+
+	a.done = true
+
+	// Cut n,a=username,\x01host=...\x01auth=...
+	// into
+	//   n
+	//   a=username
+	//   \x01host=...\x01auth=...\x01\x01
+	parts := bytes.SplitN(response, []byte{','}, 3)
+	if len(parts) != 3 {
+		return fail("Invalid response")
+	}
+	if !bytes.Equal(parts[0], []byte{'n'}) {
+		return fail("Invalid response, missing 'n'")
+	}
+	opts := OAuthBearerOptions{}
+	if !bytes.HasPrefix(parts[1], []byte("a=")) {
+		return fail("Invalid response, missing 'a'")
+	}
+	opts.Username = string(bytes.TrimPrefix(parts[1], []byte("a=")))
+
+	// Cut \x01host=...\x01auth=...\x01\x01
+	// into
+	//   *empty*
+	//   host=...
+	//   auth=...
+	//   *empty*
+	//
+	// Note that this code does not do a lot of checks to make sure the input
+	// follows the exact format specified by RFC.
+	params := bytes.Split(parts[2], []byte{0x01})
+	for _, p := range params {
+		// Skip empty fields (one at start and end).
+		if len(p) == 0 {
+			continue
+		}
+
+		pParts := bytes.SplitN(p, []byte{'='}, 2)
+		if len(pParts) != 2 {
+			return fail("Invalid response, missing '='")
+		}
+
+		switch string(pParts[0]) {
+		case "host":
+			opts.Host = string(pParts[1])
+		case "port":
+			port, err := strconv.ParseUint(string(pParts[1]), 10, 16)
+			if err != nil {
+				return fail("Invalid response, malformed 'port' value")
+			}
+			opts.Port = int(port)
+		case "auth":
+			const prefix = "bearer "
```

Comment:
Should we support other token types?

Reply:
Doesn't seem like it.

Reply:
RFC 7628 describes OAUTHBEARER mechanism as one using Bearer tokens. https://tools.ietf.org/html/rfc7628#section-3

No support for other token types is defined (besides OAUTH10B but this is a different mechanism).